### PR TITLE
First implementation of activeSlotCoefficient

### PR DESCRIPTION
**Pull request by MarcoOl94**
## Description
<Description of the pull request>

## Jira Ticket
[<Ticket Number>](<Ticket Url>)

## Changes
<Changes made> 

## Breaking Changes
<Please remember to mention anyone that may be affected by these changes so that they are aware of this PR>
- <@...> 

## Checks
- [x] Project Builds
- [x] Project passes tests and checks
- [ ] Updated documentation accordingly
- [ ] Breaking changes have been correctly tagged and notified

### CODE REVIEW

File: sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package io.horizen.fork
+
+case class ActiveSlotCoefficientFork(
+                                    activeSlotCoefficient: Double = -1
+                                    ) extends OptionalSidechainFork
```
Comment (paolocappelletti) [inline sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala:5]: let's add a check that  activeSlotCoefficient (if defined) is > 0

File: sdk/src/main/scala/io/horizen/forge/AbstractForger.scala [hunk 2]
```diff
@@ -46,6 +46,7 @@ abstract class AbstractForger[
   type View = CurrentView[HIS, MS, VL, MP]
 
   private val restApiTimeoutDuration: FiniteDuration = settings.sparkzSettings.restApi.timeout
+  private var slotsDurationInSeconds = ConsensusParamsUtil.getConsensusSecondsInSlotsPerEpoch(params.sidechainGenesisBlockTimestamp, timeProvider.time() / 1000) * 1000
```
Comment (paolocappelletti) [inline sdk/src/main/scala/io/horizen/forge/AbstractForger.scala:49]: looks to me not 100% safe to initialize this here (the time the class is instantiated and the time startTimer is called for the first time could differ?)
we may initialize this to null and in startTimer add something like this:
if (slotsDurationInSeconds == null) {
   slotsDurationInSeconds  =  ConsensusParamsUtil.getConsensusSecondsInSlotsPerEpoch(params.sidechainGenesisBlockTimestamp, timeProvider.time() / 1000) * 1000
}

**Review by paolocappelletti — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/io/horizen/forge/AbstractForger.scala [hunk 3]
```diff
@@ -135,7 +136,18 @@ abstract class AbstractForger[
     val currentTime: Long = timeProvider.time() / 1000
     val epochAndSlot = TimeToEpochUtils.timestampToEpochAndSlot(params.sidechainGenesisBlockTimestamp, currentTime)
     log.info(s"Send TryForgeNextBlockForEpochAndSlot message with epoch and slot $epochAndSlot")
-    tryToCreateBlockForEpochAndSlot(epochAndSlot.epochNumber, epochAndSlot.slotNumber, None, Seq())  }
+    tryToCreateBlockForEpochAndSlot(epochAndSlot.epochNumber, epochAndSlot.slotNumber, None, Seq())
+    RecalculateSlotDuration()
+  }
+
+  protected def RecalculateSlotDuration(): Unit = {
```
[the remark this answers is not in the document]
Comment (MarcoOl94) [reply sdk/src/main/scala/io/horizen/forge/AbstractForger.scala:143]: Done

File: sdk/src/main/scala/io/horizen/forge/AbstractForger.scala [hunk 4]
```diff
@@ -135,7 +136,18 @@ abstract class AbstractForger[
...
+    val consensusSecondsInSlot = ConsensusParamsUtil.getConsensusSecondsInSlotsPerEpoch(params.sidechainGenesisBlockTimestamp, timeProvider.time() / 1000) * 1000
+    if (consensusSecondsInSlot != slotsDurationInSeconds) {
+      slotsDurationInSeconds = consensusSecondsInSlot
+      stopTimer()
```
[the remark this answers is not in the document]
Comment (MarcoOl94) [reply sdk/src/main/scala/io/horizen/forge/AbstractForger.scala:147]: Done

File: sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package io.horizen.fork
+
+case class ActiveSlotCoefficientFork(
+                                    activeSlotCoefficient: Double = -1
```
[the remark this answers is not in the document]
Comment (MarcoOl94) [reply sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala:4]: Done

File: sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala (hunk 1 above)
Comment (MarcoOl94) [reply sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala:5]: Done

File: sdk/src/main/scala/io/horizen/forge/AbstractForger.scala (hunk 2 above)
Comment (MarcoOl94) [reply sdk/src/main/scala/io/horizen/forge/AbstractForger.scala:49]: Done

File: sdk/src/main/scala/io/horizen/consensus/package.scala [hunk 6]
```diff
@@ -93,17 +93,25 @@ package object consensus {
...
     val requiredStakePercentage: BigDecimal = vrfOutputToRequiredStakePercentage(vrfOutput, stakePercentageFork)
     val actualStakePercentage: BigDecimal = new BigDecimal(actualStake).divide(new BigDecimal(totalStake), stakeConsensusDivideMathContext)
 
-    requiredStakePercentage.compareTo(actualStakePercentage) match {
-      case -1 => true //required percentage is less than actual
-      case  0 => true //required percentage is equal to actual
-      case  _ => false //any other case
+    if (activeSlotCoefficient >= 0) {
+      val actualStakePercentageWithActiveSlotCoefficient = 1 - Math.pow( 1-activeSlotCoefficient, actualStakePercentage.doubleValue())
+      //System.out.println("Actual stake %: "+actualStakePercentage+" Actual stake % with f: "+new BigDecimal(actualStakePercentageWithActiveSlotCoefficient)+" Required stake %: "+requiredStakePercentage)
```
Comment (davehorizen) [inline sdk/src/main/scala/io/horizen/consensus/package.scala:102]: This line can be removed

File: sdk/src/main/scala/io/horizen/consensus/package.scala [hunk 7]
```diff
@@ -93,17 +93,25 @@ package object consensus {
...
+  def vrfProofCheckAgainstStake(vrfOutput: VrfOutput, actualStake: Long, totalStake: Long, stakePercentageFork: Boolean, activeSlotCoefficient: Double): Boolean = {
     val requiredStakePercentage: BigDecimal = vrfOutputToRequiredStakePercentage(vrfOutput, stakePercentageFork)
     val actualStakePercentage: BigDecimal = new BigDecimal(actualStake).divide(new BigDecimal(totalStake), stakeConsensusDivideMathContext)
 
-    requiredStakePercentage.compareTo(actualStakePercentage) match {
-      case -1 => true //required percentage is less than actual
-      case  0 => true //required percentage is equal to actual
-      case  _ => false //any other case
+    if (activeSlotCoefficient >= 0) {
+      val actualStakePercentageWithActiveSlotCoefficient = 1 - Math.pow( 1-activeSlotCoefficient, actualStakePercentage.doubleValue())
```
Comment (davehorizen) [inline sdk/src/main/scala/io/horizen/consensus/package.scala:101]: Maybe I'm missing some kind of info but shouldn't it be the other way around? That is, the `actualStakePercentage` as base number of the `Math.pow` function and the `1-activeSlotCoefficient` as exponent. In this way we will have a value of `actualStakePercentage` corrected by the exponent of domain _[0:1]_

**Review by paolocappelletti — APPROVED** [review]

File: sdk/src/main/scala/io/horizen/consensus/package.scala (hunk 6 above)
Comment (MarcoOl94) [reply sdk/src/main/scala/io/horizen/consensus/package.scala:102]: Done

File: sdk/src/main/scala/io/horizen/consensus/package.scala (hunk 7 above)
Comment (MarcoOl94) [reply sdk/src/main/scala/io/horizen/consensus/package.scala:101]: The formula was changed accordingly to the analysis made by the Crypto Team. You can find the results here: https://horizenlabs.atlassian.net/browse/CRYPTO-479
Please tell me if you still have concerns.

File: sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package io.horizen.fork
+
+case class ActiveSlotCoefficientFork(
```
Comment (davehorizen) [inline sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala:3]: Must the slot coefficient have a maximum ceiling value of 1 (100% of slots)?

Comment (MarcoOl94) [reply sdk/src/main/scala/io/horizen/fork/ActiveSlotCoefficientFork.scala:3]: Yes good point.

**Review by davehorizen — APPROVED** [review]

**Pull request merged**